### ISSUE TITLE
Deploy Report Stream OpenAPI docs to Staging Environment

### ISSUE THREAD

**Issue by JFU-NAVA-PBC** [CDCgov/prime-reportstream#10262]
### User Story

As a report stream API consumer / developer, I want report stream API's OpenAPI spec deployed to staging env ( as a proof of concept for official hosting of Report Stream API docs)

As a POC, we can deploy part of the API docs generated from ApiKeysFunctions.kt (See #9285 for details)

### Description/Use Case

Deploy our API docs to staging as a POC of where is the best place to host our API docs

### Risks/Impacts/Considerations

No impact on functional code

### Dev Notes

As a POC, we can deploy part of the API docs generated from ApiKeysFunctions.kt (See #9285 for details)

### Acceptance Criteria

- [ ] Existing generated API documentation (ApiKeysFunctions.kt (yaml)) deployed to staging env, and available for team evaluation.
- [ ] The "try it out" feature of swagger works for all endpoints
- [ ] Adjusted api documentation based on findings - new story https://github.com/CDCgov/prime-reportstream/issues/10224 TBD
- [ ] A feature switch to turn the API docs end point ON/OFF on PROD/STAGING

Note about AC:

Azure portal has means to turn on and off a web site hosted from blob container
Only deploy to staging

For completeness of info, Azure has API Management which also deal with OpenAPI docs for the API hosted, however, it is not sure if that is applicable to RS kotlin based API implementation:

https://learn.microsoft.com/en-us/azure/azure-functions/functions-openapi-definition#create-the-api-management-instance

Note, it turns out that the "feature switch" is already available for Azure Functions, see below quote, i.e. leverage Azure Functions security policies, so no need to invent the wheel:

**The recommended way to disable a function is with an app setting in the format AzureWebJobs. <FUNCTION_NAME>. Disabled set to true . You can create and modify this application setting in several ways, including by using the Azure CLI and from your function's Overview tab in the Azure portal.**

**Comment by JFU-NAVA-PBC** [comment]
FYI, 

Tried to host RS swagger UI docs from a azure blob storage (local host on port 10000, azurite), and it works.

There is some twisted situation: the API are on localhost:7071, and the site (swagger UI API docs) is on localhost:10000, so there are some CORS settings on both storage services and API services to tweak.

For deployment to PROD, and STAGGING

We probably need more scrutiny on security side:

We probably only want to deploy the API docs to STAGGING, because:

Basically, we need to try to avoid hosting live API docs on PROD, because we never know the swagger stuff is clear of vulnerabilities, and Live API docs basically is still for API consumer to learn the API, and there is no need to mess with PROD data.

Also, get approval from CDC stake holder for live API docs deployment.

**Comment by JFU-NAVA-PBC** [comment]
FYI,

1. Hosted swagger UI with generated api.yaml (see #9285) on a local setup with blob storage on localhost:10000 (need some configuration changes to relax CORS and access level), drag and drop swagger UI dist 5.1.0 with api.yaml onto azurite explorer blob storage container - created a container "site" to hold the swagger stuff
2. Point browser (chrome used here) to swagger UI page: http://127.0.0.1:10000/devstoreaccount1/site/swagger-ui-5.1.0/dist/index.html
3. From the swagger UI page go to e.g. GET and click "try it" and put "simple_report" as the arg name, then click "Execute", and a 200 OK code and keys payload will be rendered as shown:
4. 
http://127.0.0.1:10000/devstoreaccount1/site/swagger-ui-5.1.0/dist/index.html

Note, to avoid the normal browser hit the CORS error (see screen shot), I temporarily disabled chrome's CORS checking: start chrome like this: (jfu@DESKTOP-NU14HVQ:~/TEMP$ google-chrome --disable-web-security --user-data-dir="~/TEMP")

This will make sure API calls into localhost:7071 from html page loaded from localhost:10000 will not be rejected (by the chrome browser), this is just for POC.

Under normal chrome setup, execute GET will run into error:

![image.png](https://images.zenhubusercontent.com/648b4c8bc38d587baa4924d6/18081233-92e6-4bf0-b279-3e946e735d19) 

the details of the error:

Access to fetch at 'http://localhost:7071/v1/settings/organizations/simple_report/public-keys' from origin 'http://127.0.0.1:10000' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

**Comment by JFU-NAVA-PBC** [comment]
@bishoyayoub my estimate is 3 points

**Comment by JFU-NAVA-PBC** [comment]
Hosting ReportStream APIDocs3.0 Swagger UI From Azure Blob Container

(Only provisioned on staging for security)

Steps:

A. Manual Provision of APIDocs (yaml) and Swagger UI content

B. Github Action: Upload APIDocs (yaml) and Swagger UI content to the container "swaggeruiapidocs"

**Comment by JFU-NAVA-PBC** [comment]
@arnejduranovic @bishoyayoub 

Status:

 I tried to do a demo1 deploy to test the github action changes before going to staging, following some documents (may be no longer current) at: 

https://github.com/CDCgov/prime-reportstream/blob/master/prime-router/docs/docs-deprecated/environment-provisioning.md

Reached out to dev op through slack:

`If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
Run terraform validate
╷
│ Error: Missing required argument
│ 
│   on main.tf line 215, in module "metabase":
│  215: module "metabase" ***
│ 
│ The argument "sendgrid_password" is required, but no definition was found.
╵
Error: Process completed with exit code 1.
Run sleep 300;
ERROR: (ResourceNotFound) The Resource 'Microsoft.DBforPostgreSQL/servers/pdhdemo1-pgsql' under resource group 'prime-data-hub-demo1' was not found. For more details please go to https://aka.ms/ARMResourceNotFoundFix
Code: ResourceNotFound
Message: The Resource 'Microsoft.DBforPostgreSQL/servers/pdhdemo1-pgsql' under resource group 'prime-data-hub-demo1' was not found. For more details please go to https://aka.ms/ARMResourceNotFoundFix
Error: Process completed with exit code 3.`

DevOp will look into it:

Joseph:

quote slack:

"Josiah Siegel:
  3 hours ago
looks like that pass is missing for the demo envs. i should be able to fix that today
:thank-you:
"

**Comment by JFU-NAVA-PBC** [comment]
for PROD review:

checked the staging swagger ui contents after each deploy, and noticed that the github action uploading swagger ui is doing the job:

this is the timestamp after todays deployment:

[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBeW9OQVE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--4a33830e7f6e39cd96ef248d897049e7dc6900d1/Screenshot%202023-10-05%20at%2012.30.43%20PM.png)
